demos/bluetooth/mesh/thingy52/thingy52_music_cli/src/model_handler.c

```diff
@@ -0,0 +1,355 @@
+				    sizeof(mic_data.pdm_buf_b) /
+					    sizeof(mic_data.pdm_buf_b[0]));
+
+		if (!mic_data.fft_result_ready) {
+			arm_rfft_q15(&S15, (q15_t *)mic_data.pdm_buf_a,
+				     mic_data.fft_buf);
+			arm_cmplx_mag_q15(mic_data.fft_buf, mic_data.spectrum,
+					  SPECTRUM_SIZE);
+
+			mic_data.fft_result_ready = true;
+		}
+	} else if (p_evt->buffer_requested &&
+		   p_evt->buffer_released == mic_data.pdm_buf_b) {
+		nrfx_pdm_buffer_set(mic_data.pdm_buf_a,
+				    sizeof(mic_data.pdm_buf_a) /
+					    sizeof(mic_data.pdm_buf_a[0]));
+
+		if (!mic_data.fft_result_ready) {
+			arm_rfft_q15(&S15, (q15_t *)mic_data.pdm_buf_b,
+				     mic_data.fft_buf);
+			arm_cmplx_mag_q15(mic_data.fft_buf, mic_data.spectrum,
+					  SPECTRUM_SIZE);
+
+			mic_data.fft_result_ready = true;
+		}
+	}
+}
+
+static void send_color_helper(int iterator, uint16_t value)
+{
+	int err = 0;
+
+	static struct bt_mesh_thingy52_rgb_msg msg = {
+		.ttl = 0,
+		.delay = 0xFFFF,
+		.speaker_on = false,
+	};
+
+	/* Should match NUM_FREQ_BANDS */
+	if (iterator == 0) {
+		msg.color.red = 0;
+		msg.color.green = 0;
+		msg.color.blue = value;
+		err |= bt_mesh_thingy52_cli_rgb_set(&thingy52_cli[iterator],
+						    NULL, &msg);
+	} else if (iterator == 1) {
+		msg.color.red = 0;
+		msg.color.green = value;
+		msg.color.blue = 0;
+		err |= bt_mesh_thingy52_cli_rgb_set(&thingy52_cli[iterator],
+						    NULL, &msg);
+	} else if (iterator == 2) {
+		msg.color.red = value;
+		msg.color.green = 0;
+		msg.color.blue = 0;
+		err |= bt_mesh_thingy52_cli_rgb_set(&thingy52_cli[iterator],
+						    NULL, &msg);
+	} else {
+		printk("Unknown iterator\n");
+	}
+
+	if (err) {
+		printk("Error sending RGB set message\n");
+	}
+}
+
+/* Values in the spectrum array are used as raw values (not q15) */
+static void microphone_work_handler(struct k_work *work)
+{
+	static uint16_t freq_sum[NUM_FREQ_BANDS];
+	static uint16_t prev_freq_sum[NUM_FREQ_BANDS];
+	static uint16_t total_sum;
+	static uint16_t checker;
+	static uint8_t divider;
+	static uint16_t adj_freq_sum;
+
+	if (mic_data.fft_result_ready) {
+		int i;
+
+		for (i = 0; i < NUM_FREQ_BANDS; i++) {
+			freq_sum[i] = 0;
+		}
+
+		/* Should create the amount of for-loops
+		 * that macthes NUM_FREQ_BANDS
+		 */
+		for (i = 0; i <= 80; i++) {
+			if (mic_data.spectrum[i] > 100) {
+				freq_sum[0] += mic_data.spectrum[i];
+			}
+		}
+
+		for (i = 81; i <= 115; i++) {
+			if (mic_data.spectrum[i] > 100) {
+				freq_sum[1] += mic_data.spectrum[i];
+			}
+		}
+
+		for (i = 116; i <= 255; i++) {
+			if (mic_data.spectrum[i] > 100) {
+				freq_sum[2] += mic_data.spectrum[i];
+			}
+		}
+
+		total_sum = freq_sum[0] + freq_sum[1] + freq_sum[2];
+
+		/* A basic way of handling volume changes */
+		if (total_sum) {
+			checker = total_sum / (4 + total_sum / 5000);
+			divider = 6 + total_sum / 1000;
+		}
+
+		for (i = 0; i < NUM_FREQ_BANDS; i++) {
+			if (freq_sum[i]) {
+				if (freq_sum[i] - prev_freq_sum[i] > checker) {
```

Comment:
I'm not sure what this does. If it's a threshold for change, the left hand side should be surrounded by `abs()`

Reply:
Not using abs() because only the _increasing_ volume change desired

Reply:
Made a comment about this